modules/dists/BlockDist.chpl

```diff
@@ -1536,12 +1536,17 @@ where this.sparseLayoutType == unmanaged DefaultDist &&
 }
 
 // Block1 <=> Block2 
-proc BlockArr.doiSwap(arr) {
-  coforall (locarr1, locarr2) in zip(this.locArr, arr.locArr) {
-    on locarr1 {
-      locarr1.myElems <=> locarr2.myElems;
-      locarr1.locRAD <=> locarr2.locRAD;
+proc BlockArr.doiOptimizedSwap(other) {
+  if(this.dom.dist.dsiEqualDMaps(other.dom.dist)) {
+    coforall (locarr1, locarr2) in zip(this.locArr, other.locArr) {
+      on locarr1 {
+        locarr1.myElems <=> locarr2.myElems;
+        locarr1.locRAD <=> locarr2.locRAD;
```

Comment:
@mcdobe100 / @e-kayrakli : Looking at this code tonight, it appears to me that if we have two Block-distributed arrays with equivalent distributions, that we will attempt to swap them even if they have different bounds.  E.g., `{1..100} dmapped Block({1..100})` and `{0..99} dmapped Block({1..100})`.  I think that this will pass the dsiEqualDMaps() check, but then fall apart when we try to swap local blocks that happen to be different sizes (when it should work and probably fall back on the more expensive default solution).  Do you agree, or am I missing something else?  Thanks.

Reply:
I believe this is handled by default rectangular's swap right below at https://github.com/chapel-lang/chapel/pull/16246/files/ea28d02ff601bc0dd362286cba0367827264417b

You can think of the optimization having two levels. In your scenario, the "outer" block array will do optimized swap as it won't be iterated element by element and will just swap the `myElems` arrays. However, their swap will not use the optimized swap and do the one-by-one iteration. Which I expect to fail with `--no-fast` because of uneven zippering and have undefined behavior with `--fast`

Reply:
But (unless I'm misunderstanding something in your explanation), I don't think this is right:  It should be legal to swap a block-distributed array over 1..100 with one over 0..99 since they're both the same size and shape.  So when I distribute these over two locales (say) using the same 1..100 bounding box, locale 0 will own 50 elements of the first and 51 elements of the second, so the local array swap will fail, but it shouldn't.  It just shouldn't be trying to rely on a local array swap and fall back to the (very expensive) element-by-element swap.

Reply:
Yep, you are right. Working on a fix to improve the checks

Reply:
OK, cool.  I think the safest/easiest fix would be to just ensure that the domains also have the same indices (?).

Reply:
https://github.com/chapel-lang/chapel/pull/16620 fixes this and some other unrelated swap issues.

Reply:
Thanks, Engin!